app/middleware/receiveRawData.js

```diff
@@ -9,65 +9,82 @@ import {
 
-export default function rawDataMiddleware({ getState }) {
-  return (dispatch) => (action) => {
+export default function rawDataMiddleware(store) {
+  return (next) => (action) => {
     if (action.type !== RECEIVE_RAW_DATA) {
-      return dispatch(action);
+      return next(action);
     }
 
-    const storeUpdates = {
+    const dateFormat = store.getState().dateFormat;
+    const newState = {
       dataFields: [],
       dataStatus: {},
       parsedData: [],
       transformedData: {},
     };
+
+    // Case 1: Empty text area
     if (!action.data.length) {
-      // Case 1: Empty text area
-      dispatch(actionTrigger(CLEAR_ERROR));
-    } else {
-      const parserResult = parseRawData(Papa, action.data);
+      return dispatchNewState(next, action, newState);
+    }
+
+    const [data, fields, errors] = parseRawData(Papa, action.data);
+    const transformedDataMap = transformParsedData(
+      data,
+      fields,
+      dateFormat
+    );
+    const combinedErrors = errors.concat(
+      validateTransformedData(transformedDataMap)
+    );
 
-      // Case 2: CSV parsing error(s)
-      if (parserResult[2].length) {
-        storeUpdates.dataStatus = {
+    // Case 2: CSV parsing error(s)
+    if (combinedErrors.length) {
+      return dispatchNewState(next, action, Object.assign(newState, {
+        dataStatus: {
           status: 'error',
-          message: parserResult[2].join('; '),
-        };
-        dispatch(actionTrigger(RECEIVE_ERROR, 'e001'));
-      } else {
-        // Case 3: CSV parsing success
-        storeUpdates.dataFields = parserResult[1];
-        storeUpdates.dataStatus = {
-          status: 'success',
-          message: 'Data input successful',
-        };
-        storeUpdates.parsedData = parserResult[0];
-        storeUpdates.transformedData = transformParsedData(
-          parserResult[0],
-          parserResult[1],
-          getState().dateFormat
-        );
-        dispatch(actionTrigger(CLEAR_ERROR));
-      }
+          message: combinedErrors.join('; '),
+        },
+      }));
     }
-    // Empty for Case 1 and Case 2, array of fields for Case 3
-    dispatch(actionTrigger(
-      PARSE_DATA_FIELDS, storeUpdates.dataFields, action.src));
 
-    // Empty for Case 1, error for Case 2, success for Case 3
-    dispatch(actionTrigger(
-      PARSE_DATA_STATUS, storeUpdates.dataStatus, action.src));
+    // Case 3: CSV parsing success
+    return dispatchNewState(next, action, Object.assign(newState, {
+      dataFields: fields,
+      dataStatus: {
+        status: 'success',
+        message: 'Data input successful',
+      },
+      parsedData: data,
+      transformedData: transformedDataMap,
+    }));
+  };
+}
 
-    // Empty for Case 1 and Case 2, array of data for Case 3
-    dispatch(actionTrigger(
-      PARSE_RAW_DATA, storeUpdates.parsedData, action.src));
+function dispatchNewState(dispatch, action, state) {
```

Comment:
This function is a helper for this file and only makes sense within it's context, because the actions it dispatches are only related to receiving data. I will rename it something less generic to reflect that.